R/brreg-utils.R

```diff
@@ -82,6 +87,54 @@ parse_brreg_entity_single <- function(p) {
   )
 }
 
+#' parse_brreg_roles
+#' @param parsed Output of `parse_response()`
+#' @inheritParams get_roles
+#' @noRd
+parse_brreg_roles <- function(parsed, entity) {
+  p <- lapply(parsed$rollegrupper, function(x) x$roller[[1]])
+  dl <- purrr::map2(p, entity, parse_brreg_roles_single)
+  dl_entities <- lapply(dl, function(x) {
+    if ('organisasjonsnummer' %in% names(x)) x else NULL
+  })
+  dl_persons <- lapply(dl, function(x) {
+    if (!'organisasjonsnummer' %in% names(x)) x else NULL
+  })
+  list(persons = do.call('rbind', dl_persons),
+       entities = do.call('rbind', dl_entities))
+}
+
+#' parse_brreg_roles_single
+#' @noRd
+parse_brreg_roles_single <- function(p, entity) {
+  if (any(grepl('person', names(p)))) {
+    data.frame(
+      aktornummer = entity,
+      type_kode = p$type$kode,
+      type_beskrivelse = p$type$beskrivelse,
+      navn = paste(p$person$navn$fornavn,
+                   p$person$navn$etternavn),
+      fodselsdato = p$person$fodselsdato,
+      er_doed = p$person$erDoed,
+      fratraadt = p$fratraadt,
+      rekkefolge = p$rekkefolge
+    )
+  } else {
+    data.frame(
+      aktornummer = entity,
+      type_kode = p$type$kode,
+      type_beskrivelse = p$type$beskrivelse,
+      organisasjonsnummer = p$enhet$organisasjonsnummer,
+      navn = paste(unlist(p$enhet$navn), collapse = ' '),
+      organisasjonsform_kode = p$enhet$organisasjonsform$kode,
+      organisasjonsform_beskrivelse = p$enhet$organisasjonsform$beskrivelse,
+      er_slettet = p$enhet$erSlettet,
+      fratraadt = p$fratraadt,
+      rekkefolge = p$rekkefolge
+    )
+  }
+}
```

Comment:
Should the utility function `null2na()` be used here as well, or are we certain that the fields will be available for all requests?

Reply:
Good question. I haven't tested this well enough to know for sure. The response here is somewhat different than the one of from `/enheter`, and it helps that we are treating individuals and entities separately. But as a "safety measure", I have added `null2na()` to most of these list items. I would assume that information on role type (`type_kode`), name and organisation number is always present, but for the rest of the items I guess it is possible that the information varies.